### PR TITLE
Implement autocomplete's Login

### PR DESCRIPTION
**Pull request by mshin-wolvic**
- WolvicPasswordManagerClient for Login save
- WolvicPasswordStorageBackend for Login DB
- WolvicAutofillClient for autofill username/password
- WolvicSignClient is dummy class to avoid crash issue
- Changed components/autofill not to show suggestions of login whenever text field is changed.
- Own Wolvic.pak for resources instead of content_shell.pak to include resources in components
- WolvicRenderFrameObserver for binding Autofill Agent's mojo

### CODE REVIEW

File: wolvic/browser/autocomplete/wolvic_password_store_backend.cc [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+      GetAndEraseCallback<password_manager::LoginsOrErrorReply>(reply_id);
+  if (!reply.has_value()) {
+    LOG(ERROR) << "OnCompleteWithLogins does not have reply in map by id["
+               << reply_id << "]";
+    return;
+  }
+  main_task_runner_->PostTask(
+      FROM_HERE,
+      base::BindOnce(std::move(*reply), std::move(passwords)));
+  // std::move(*reply).Run(std::move(passwords));
```
Comment (zakharvoit) [inline wolvic/browser/autocomplete/wolvic_password_store_backend.cc]: nit: Remove commented out code

**Review by zakharvoit — APPROVED** [review]
Didn't look too closely due to the amount of changes but overall design looks good.

File: components/autofill/content/renderer/autofill_agent.cc [hunk 2]
```diff
@@ -1006,6 +1006,15 @@ void AutofillAgent::ShowSuggestions(
...
       return;
     }
+
+    // [Wolvic] Do not show suggestions if the field is not empty since we have
+    // a different UX.
+    if (trigger_source !=
+        AutofillSuggestionTriggerSource::kFormControlElementClicked ||
+        (!element.Value().IsEmpty() && element.UserHasEditedTheField())) {
+      return;
+    }
```
Comment (svillar) [inline components/autofill/content/renderer/autofill_agent.cc:1016]: Isn't it possible to tell chromium not to use the autofill component instead?

File: components/autofill/content/renderer/password_autofill_agent.cc [hunk 3]
```diff
@@ -757,8 +757,12 @@ bool PasswordAutofillAgent::TextDidChangeInTextField(
     iter->second.password_was_edited_last = false;
   }
 
+  // [Wolvic] Do not show suggestions whenever text filed is changed since
+  // we has a different UX.
+  return false;
+
```
Comment (svillar) [inline components/autofill/content/renderer/password_autofill_agent.cc:763]: Ditto

File: wolvic/BUILD.gn [hunk 4]
```diff
@@ -108,6 +121,61 @@ shared_library("libcontent_native") {
   configs -= [ "//build/config/android:hide_all_but_jni_onload" ]
 }
 
+repack("pak") {
```
Comment (svillar) [inline wolvic/BUILD.gn:124]: why is this needed?

File: wolvic/wolvic_content_main_delegate.cc [hunk 5]
```diff
@@ -243,14 +243,14 @@ void WolvicContentMainDelegate::InitializeResourceBundle() {
     pak_region = global_descriptors->GetRegion(kShellPakDescriptor);
   } else {
     pak_fd =
-        base::android::OpenApkAsset("assets/content_shell.pak", &pak_region);
+        base::android::OpenApkAsset("assets/wolvic.pak", &pak_region);
```
Comment (svillar) [inline wolvic/wolvic_content_main_delegate.cc:246]: Can we have these changes  in a separate commit of this same PR?

File: wolvic/java/org/chromium/wolvic/AutofillManager.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        return mBridge.isPasswordManagerEnabled(windowAndroid.getContext().get());
+    }
+
+    @CalledByNative
+    public void createUsernameArray(int arraySize) {
+        if (arraySize > 0) {
+            mUserNames = new String[arraySize];
+        } else {
+            mUserNames = null;
+        }
```
Comment (svillar) [inline wolvic/java/org/chromium/wolvic/AutofillManager.java]: `mUserNames = arraySize > 0 ? new String[arraySize] : null`

File: wolvic/java/org/chromium/wolvic/AutofillManager.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @CalledByNative
+    public void addUsername(int idx, String username) {
+        mUserNames[idx] = username;
```
Comment (svillar) [inline wolvic/java/org/chromium/wolvic/AutofillManager.java:79]: are we sure that we'd always get a valid `idx` here? I wouldn't like to get an out of bounds exception

File: wolvic/browser/autocomplete/wolvic_signin_client.h [hunk 8]
```diff
@@ -0,0 +1,79 @@
+// Copyright 2024 The Chromium Authors
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef WOLVIC_BROWSER_AUTOCOMPLETE_WOLVIC_WOLVIC_SIGNIN_CLIENT_H_
```
Comment (svillar) [inline wolvic/browser/autocomplete/wolvic_signin_client.h]: Shouldn't it be `WOLVIC_BROWSER_AUTOCOMPLETE_WOLVIC_SIGNIN_CLIENT_H_`

File: wolvic/browser/autocomplete/wolvic_autofill_client.cc [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+  return nullptr;
+}
+
+autofill::payments::PaymentsClient* WolvicAutofillClient::GetPaymentsClient() {
+  return nullptr;
+}
+
+autofill::StrikeDatabase* WolvicAutofillClient::GetStrikeDatabase() {
+  return nullptr;
+}
```
Comment (svillar) [inline wolvic/browser/autocomplete/wolvic_autofill_client.cc:122]: What about adding all those `return nullptr` in the header file? Like `GetX() override { return nullptr; }` to simplify a bit this file? Same for the other similar getters including the ones with no code inside.

File: wolvic/browser/autocomplete/wolvic_autofill_client.cc [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+void WolvicAutofillClient::HideTouchToFillCreditCard() {
+  // Touch To Fill is not supported yet.
+  NOTREACHED();
+}
+
+void WolvicAutofillClient::OnLoginSelected(JNIEnv* env, jint index) {
+  if (!delegate_)
+    return;
+ 
+  if (index < 0 ||static_cast<size_t>(index) >= suggestions_.size()) {
```
Comment (svillar) [inline wolvic/browser/autocomplete/wolvic_autofill_client.cc]: blank space after `||`

File: wolvic/browser/autocomplete/wolvic_password_manager_client.cc [hunk 11]
```diff
@@ -0,0 +1,482 @@
...
+    return true;
+  return false;
+}
+
+net::CertStatus WolvicPasswordManagerClient::GetMainFrameCertStatus() const {
+  content::NavigationEntry* entry =
+      web_contents()->GetController().GetLastCommittedEntry();
+  if (!entry)
+    return 0;
+  return entry->GetSSL().cert_status;
```
Comment (svillar) [inline wolvic/browser/autocomplete/wolvic_password_manager_client.cc]: `return entry ? entry->GetSSL().cert_status : 0`

File: wolvic/browser/autocomplete/wolvic_password_manager_client.cc [hunk 12]
```diff
@@ -0,0 +1,482 @@
...
+  DCHECK(web_contents());
+
+  content::NavigationEntry* entry =
+      web_contents()->GetController().GetVisibleEntry();
+  if (!entry)
+    return false;
+  int http_status_code = entry->GetHttpStatusCode();
+  if (http_status_code >= 400 && http_status_code < 600)
+    return true;
+  return false;
```
Comment (svillar) [inline wolvic/browser/autocomplete/wolvic_password_manager_client.cc]: Just `return http_status_code >= 400 && http_status_code < 600;`

**Review by svillar — CHANGES_REQUESTED** [review]
Amazing piece of work. I left some comments inline. Apart from that I have some questions:
* I'm a bit scared of adding so much amount of code (even knowing that there are many dummy implementations). Can't we just inherit from another existing clients?
* I don't have a clear idea of how all the actors interact in this whole process, it'd be nice if you could write some brief document explaining how all this work.
* What's the "crash issue" that forces us to implement the `WolvicSignClient`?

File: components/autofill/content/renderer/autofill_agent.cc (hunk 2 above)
Comment (mshin-wolvic) [reply components/autofill/content/renderer/autofill_agent.cc:1016]: Do you mean that turning on/off the autofill component according to this condition?
If yes, it would be possible, but this code runs on the renderer process, and the addition of a Mojo message is required to send a message to the browser process. We need to touch up the origin chromium code a bit more. But we would want to keep the origin code to a minimum for rebase cost. Please let me know if I misinterpret your comment.

File: wolvic/BUILD.gn (hunk 4 above)
Comment (mshin-wolvic) [reply wolvic/BUILD.gn:124]: Currently we are using content_shell.pak that configured for testing and it's hard to extend other resources. This is for including //components resources in Wolvic.

File: wolvic/browser/autocomplete/wolvic_signin_client.h (hunk 8 above)
Comment (mshin-wolvic) [reply wolvic/browser/autocomplete/wolvic_signin_client.h]: Oh, right! Done

File: wolvic/java/org/chromium/wolvic/AutofillManager.java (hunk 6 above)
Comment (mshin-wolvic) [reply wolvic/java/org/chromium/wolvic/AutofillManager.java]: Done

File: wolvic/java/org/chromium/wolvic/AutofillManager.java (hunk 7 above)
Comment (mshin-wolvic) [reply wolvic/java/org/chromium/wolvic/AutofillManager.java:79]: Added the safe check. Thanks!

File: wolvic/wolvic_content_main_delegate.cc (hunk 5 above)
Comment (mshin-wolvic) [reply wolvic/wolvic_content_main_delegate.cc:246]: Done

File: wolvic/browser/autocomplete/wolvic_password_store_backend.cc (hunk 1 above)
Comment (mshin-wolvic) [reply wolvic/browser/autocomplete/wolvic_password_store_backend.cc]: Done

File: wolvic/browser/autocomplete/wolvic_autofill_client.cc (hunk 9 above)
Comment (mshin-wolvic) [reply wolvic/browser/autocomplete/wolvic_autofill_client.cc:122]: We will meet the build error since Chromium don't allow to have the implementation of overridden method in header file.

File: wolvic/browser/autocomplete/wolvic_autofill_client.cc (hunk 10 above)
Comment (mshin-wolvic) [reply wolvic/browser/autocomplete/wolvic_autofill_client.cc]: Done

File: wolvic/browser/autocomplete/wolvic_password_manager_client.cc (hunk 12 above)
Comment (mshin-wolvic) [reply wolvic/browser/autocomplete/wolvic_password_manager_client.cc]: Done

**Comment by mshin-wolvic** [comment]
> * I'm a bit scared of adding so much amount of code (even knowing that there are many dummy implementations). Can't we just inherit from another existing clients?

Unfortunately, password manager component has Stubxxx, Mockxxx for test only and there is no inheritable client.

> * I don't have a clear idea of how all the actors interact in this whole process, it'd be nice if you could write some brief document explaining how all this work.

OK, I will share the document.

> * What's the "crash issue" that forces us to implement the `WolvicSignClient`?

We don't actually use Identity or Signin of Chromium components, but PasswordManager is tightly close to them without null check. So when this is null, it crashes.

See code here.

```
void AccountFetcherService::Initialize(
...
  repeating_timer_ = std::make_unique<signin::PersistentRepeatingTimer>(
      signin_client_->GetPrefs(), AccountFetcherService::kLastUpdatePref,
```

https://source.chromium.org/chromium/chromium/src/+/main:components/signin/internal/identity_manager/account_fetcher_service.cc;l=103?q=AccountFetcherService::Initialize&ss=chromium%2Fchromium%2Fsrc

This will only go through initialization and will no longer work, but we should provide a dummy client to avoid null access.

**Comment by mshin-wolvic** [comment]
> OK, I will share the document.
> 
> > * What's the "crash issue" that forces us to implement the `WolvicSignClient`?

**Review by svillar — APPROVED** [review]
Amazing work!
